content/r-package-dev/debugging.Rmd

```diff
@@ -0,0 +1,93 @@
+---
+title: "Debugging"
+author: "David Watkins"
+slug: "debugging"
+date: "9999-04-30"
+image: "img/main/intro-icons-300px/r-logo.png"
+output: USGSmarkdowntemplates::hugoTraining
+parent: R Package Development
+weight: 1
+draft: true
+---
+```{r setup, include=FALSE, warning=FALSE, message=FALSE}
+library(knitr)
+
+knit_hooks$set(plot=function(x, options) {
+  sprintf("<img src='../%s%s-%d.%s'/ title='%s'/>", 
+          options$fig.path, options$label, options$fig.cur, options$fig.ext, options$fig.cap)
+
+})
+
+opts_chunk$set(
+  echo=TRUE,
+  fig.path="static/writing-tests/",
+  fig.width = 6,
+  fig.height = 6,
+  fig.cap = "TODO"
+)
+
+set.seed(1)
+```
+
+Debugging is an integral part of package development and maintenance.  As you write more complex code (hopefully well-functionalized), errors can often happen several functions deep.  Unlike in a script, you don't by default have access to the R variables where the error occured.  Fortunately, RStudio provides many useful tools to help.  
+
+## Lesson Objectives
+
+1. Understand and be able to use the debugging tools in R and RStudio, namely `traceback()`, `debug`, and breakpoints.  
+2. Be able to track down an error to a particular line in an unfamiliar function, and view the environment where the error occurred
+
+## Types of errors
+
+Broadly, there are three ways your code can fail:
+
+* *syntax* - Syntax errors are generally typos that prevent your code from running in the first place, such as a missing closing parenthesis.  We won't deal with these much in this material.  Rstudio's syntax highlighting features should make these relatively obvious, with a red squiggly line or 'x' icon on the offending line.
+
+* *runtime* - Runtime errors occur when the program begins to execute, but fails part way through.
+
+* *logical* - Logical errors are the most insidious, as the program runs to completion but does not produce the desired results.
+
+If you have done a good job with testing and defensive programming **(LINK TO SECTIONS HERE)**, most all logic problems should result in explicit run-time errors.     
+
+## The debugging process
+
+Firstly, you need to determine where the error occured.    Note that is only means the *line where the function failed* -- this is not necessarily the problem line of code, only where it was caught. From here, you can work backwards to the source of the problem.  For instance, a data frame could lose a column you weren't expecting during a join, but a failure doesn't occur until much later when that missing column is actually used.
```

Comment:
so if the line where the function failed isn't the place where bad things happen, why is it good to start by finding it? tell the readers